File: Tests/Tests/STPConnectAccountParamsTest.m
//
//  STPConnectAccountParamsTest.m
//  StripeiOS Tests
//
//  Created by Daniel Jackson on 1/10/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPConnectAccountParams.h"
#import "STPConnectAccountIndividualParams.h"
#import "STPConnectAccountCompanyParams.h"

@interface STPConnectAccountParams (Testing)
+ (NSString *)stringFromBusinessType:(STPConnectAccountBusinessType)businessType;
@end

@interface STPConnectAccountParamsTest : XCTestCase
@end

@implementation STPConnectAccountParamsTest

#pragma mark - STPFormEncodable Tests

- (void)testRootObjectName {
    XCTAssertEqualObjects([STPConnectAccountParams rootObjectName], @"account");
}

- (void)testBusinessType {
    STPConnectAccountIndividualParams *individual = [STPConnectAccountIndividualParams new];
    STPConnectAccountCompanyParams *company = [STPConnectAccountCompanyParams new];
    
    XCTAssertEqual([[STPConnectAccountParams alloc] initWithIndividual:individual].businessType, STPConnectAccountBusinessTypeIndividual);
    XCTAssertEqual([[STPConnectAccountParams alloc] initWithTosShownAndAccepted:YES individual:individual].businessType, STPConnectAccountBusinessTypeIndividual);

    XCTAssertEqual([[STPConnectAccountParams alloc] initWithCompany:company].businessType, STPConnectAccountBusinessTypeCompany);
    XCTAssertEqual([[STPConnectAccountParams alloc] initWithTosShownAndAccepted:YES company:company].businessType, STPConnectAccountBusinessTypeCompany);
}

- (void)testBusinessTypeString {
    XCTAssertEqualObjects(@"individual", [STPConnectAccountParams stringFromBusinessType:STPConnectAccountBusinessTypeIndividual]);
    XCTAssertEqualObjects(@"company", [STPConnectAccountParams stringFromBusinessType:STPConnectAccountBusinessTypeCompany]);
}

- (void)testPropertyNamesToFormFieldNamesMapping {
    STPConnectAccountIndividualParams *individual = [STPConnectAccountIndividualParams new];
    STPConnectAccountParams *accountParams = [[STPConnectAccountParams alloc] initWithIndividual:individual];

    NSDictionary *mapping = [STPConnectAccountParams propertyNamesToFormFieldNamesMapping];

    for (NSString *propertyName in [mapping allKeys]) {
        XCTAssertFalse([propertyName containsString:@":"]);
        XCTAssert([accountParams respondsToSelector:NSSelectorFromString(propertyName)]);
    }

    for (NSString *formFieldName in [mapping allValues]) {
        XCTAssert([formFieldName isKindOfClass:[NSString class]]);
        XCTAssert([formFieldName length] > 0);
    }

    XCTAssertEqual([[mapping allValues] count], [[NSSet setWithArray:[mapping allValues]] count]);
}

@end

File: Stripe/StripeiOSTests/STPConnectAccountParamsTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPConnectAccountParamsTest.m
//  StripeiOS Tests
//
//  Created by Daniel Jackson on 1/10/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

@testable import StripePayments

class STPConnectAccountParamsTest: XCTestCase {
    // MARK: - STPFormEncodable Tests

    func testRootObjectName() {
        XCTAssertEqual(STPConnectAccountParams.rootObjectName(), "account")
    }

    func testBusinessType() {
        let individual = STPConnectAccountIndividualParams()
        let company = STPConnectAccountCompanyParams()

        XCTAssertEqual(STPConnectAccountParams(individual: individual).businessType, .individual)
        XCTAssertEqual(STPConnectAccountParams(tosShownAndAccepted: true, individual: individual)!.businessType, .individual)

        XCTAssertEqual(STPConnectAccountParams(company: company).businessType, .company)
        XCTAssertEqual(STPConnectAccountParams(tosShownAndAccepted: true, company: company)!.businessType, .company)
    }

    func testBusinessTypeString() {
        XCTAssertEqual("individual", STPConnectAccountParams.string(from: .individual))
        XCTAssertEqual("company", STPConnectAccountParams.string(from: .company))
        XCTAssertEqual(nil, STPConnectAccountParams.string(from: .none))
    }

    func testPropertyNamesToFormFieldNamesMapping() {
        let individual = STPConnectAccountIndividualParams()
        let accountParams = STPConnectAccountParams(individual: individual)

        let mapping = STPConnectAccountParams.propertyNamesToFormFieldNamesMapping()

        for propertyName in mapping.keys {
            XCTAssertFalse(propertyName.contains(":"))
            XCTAssert(accountParams.responds(to: NSSelectorFromString(propertyName)))
        }

        for formFieldName in mapping.values {
            XCTAssert(formFieldName.count > 0)
        }

        XCTAssertEqual(mapping.values.count, Set(mapping.values).count)
    }
}
